source/websocket_decoder.c

```diff
@@ -0,0 +1,331 @@
+#define VALUE_FOR_8BYTE_EXTENDED_LENGTH 127
+
+typedef int(state_fn)(struct aws_websocket_decoder *decoder, struct aws_byte_cursor *data);
+
+/* STATE_INIT: Resets things, consumes no data */
+static int s_state_init(struct aws_websocket_decoder *decoder, struct aws_byte_cursor *data) {
+    (void)data;
+    AWS_ZERO_STRUCT(decoder->current_frame);
+    decoder->state++;
+    return AWS_OP_SUCCESS;
+}
+
+/* STATE_OPCODE_BYTE: Decode first byte of frame, which has all kinds of goodies in it. */
+static int s_state_opcode_byte(struct aws_websocket_decoder *decoder, struct aws_byte_cursor *data) {
+    if (data->len == 0) {
+        return AWS_OP_SUCCESS;
+    }
+
+    uint8_t byte = data->ptr[0];
+    aws_byte_cursor_advance(data, 1);
+
+    /* first 4 bits are all bools */
+    decoder->current_frame.fin = byte & 0x80;
+    decoder->current_frame.rsv[0] = byte & 0x40;
+    decoder->current_frame.rsv[1] = byte & 0x20;
+    decoder->current_frame.rsv[2] = byte & 0x10;
+
+    /* next 4 bits are opcode */
+    decoder->current_frame.opcode = byte & 0x0F;
+
+    /* RFC-6455 Section 5.2 Fragmentation
+     *
+     * Data frames with the FIN bit clear are considered fragmented and must be followed by
+     * 1+ CONTINUATION frames, where only the final CONTINUATION frame's FIN bit is set.
+     *
+     * Control frames may be injected in the middle of a fragmented message,
+     * but control frames may not be fragmented themselves.
+     */
+    if (aws_websocket_is_data_frame(decoder->current_frame.opcode)) {
+        bool is_continuation_frame = AWS_WEBSOCKET_OPCODE_CONTINUATION == decoder->current_frame.opcode;
+
+        if (decoder->expecting_continuation_data_frames != is_continuation_frame) {
+            return aws_raise_error(AWS_ERROR_HTTP_PARSE);
+        }
+
+        decoder->expecting_continuation_data_frames = !decoder->current_frame.fin;
+
+    } else {
+        /* Control frames themselves MUST NOT be fragmented. */
+        if (!decoder->current_frame.fin) {
+            return aws_raise_error(AWS_ERROR_HTTP_PARSE);
+        }
+    }
+
+    decoder->state++;
+    return AWS_OP_SUCCESS;
+}
+
+/* STATE_LENGTH_BEGIN: Decode byte containing length, determine if we need to decode extended length. */
+static int s_state_length_begin(struct aws_websocket_decoder *decoder, struct aws_byte_cursor *data) {
+    if (data->len == 0) {
+        return AWS_OP_SUCCESS;
+    }
+
+    uint8_t byte = data->ptr[0];
+    aws_byte_cursor_advance(data, 1);
+
+    /* first bit is a bool */
+    decoder->current_frame.masked = byte & 0x80;
+
+    /* remaining 7 bits are payload length */
+    decoder->current_frame.payload_length = byte & 0x7F;
+
+    /* If 7bit payload length has a high value, then the next few bytes contain the real payload length */
+    if (decoder->current_frame.payload_length < VALUE_FOR_2BYTE_EXTENDED_LENGTH) {
+        decoder->state += 2; /* Skip next state, which would have processed extended length */
```

Comment:
I think I'd rather see the state increments be just setting to a state for better readability:

decoder->state = AWS_WEBSOCKET_PARSER_STATE_X;

Reply:
I thought that the incrementing thing made it more flexible, if states were added/removed in the future. Every state does += 1, except BEGIN states that sometimes do += 2 to skip their CONTINUE states.